test/helpers/FormContext.spec.js

```diff
@@ -277,4 +277,209 @@ describe('FormContext', () => {
+      children: List.of(obsRecordTree),
+    });
+
+    const obsGroupRecord = new ControlRecord({
+      control: {
+        type: 'obsGroupControl',
+        label: {
+          value: 'BMI',
+          id: '3',
+        },
+        properties: {
+          addMore: true,
+        },
+        id: '3',
+        concept: {
+          name: 'BMI Data',
+        },
+      },
+      formFieldPath: 'Vitals.2/1-0/3-0',
+      children: List.of(firstObsRecord),
+    });
+
+    it('should get first obs record out of two add-more records with the given concept name',
+        () => {
+          const formContext = new FormContext(obsRecordTree);
+          const targetRecordWrapper = formContext.getFromParent('MTC, Drug prescribed days');
+          expect(targetRecordWrapper.currentRecord.formFieldPath).to.equal('Vitals.2/2-0');
+        });
+
+    it('should get first section record out of two add-more records with the given label name',
+        () => {
+          const sectionRecordTree = new ControlRecord({
+            children: List.of(firstSectionRecord, secondSectionRecord),
+          });
+
+          const formContext = new FormContext(undefined, undefined, sectionRecordTree);
+          const targetRecordWrapper = formContext.getFromParent('Section');
+          expect(targetRecordWrapper.currentRecord.formFieldPath).to.equal('Vitals.2/1-0');
+        });
+
+    it('should get obs group record with the given concept name where obs group lies ' +
+        'under a section', () => {
+      const rootRecord = new ControlRecord({
+        children: List.of(obsGroupRecord),
+      });
+
+      const formContext = new FormContext(rootRecord, undefined, obsGroupRecord);
+      const targetRecordWrapper = formContext.getFromParent('BMI Data');
+      expect(targetRecordWrapper.currentRecord.formFieldPath).to.equal('Vitals.2/1-0/3-0');
+    });
+
+    it('should log warning message if no record available for given concept name', () => {
+      const warningSpy = sinon.spy(console, 'warn');
+      const formContext = new FormContext(new ControlRecord());
+      formContext.getFromParent('BMI Data');
+      const expectedMessage = '[FormEventHandler] Control with name[BMI Data] does not exist';
+      sinon.assert.calledOnce(warningSpy.withArgs(expectedMessage));
+      warningSpy.restore();
+    });
+
+    it('should get first obs record out of two add-more records with the given id',
+    () => {
+      const formContext = new FormContext(obsRecordTree);
+      const targetRecordWrapper = formContext.getFromParentById(2);
+      expect(targetRecordWrapper.currentRecord.formFieldPath).to.equal('Vitals.2/2-0');
+    });
+
+    it('should get first section record out of two add-more records with the given id',
+      () => {
+        const sectionRecordTree = new ControlRecord({
+          children: List.of(firstSectionRecord, secondSectionRecord),
+        });
+
+        const formContext = new FormContext(undefined, undefined, sectionRecordTree);
```

Comment:
when will we have such a scenario, where the root will be undefined?

Reply:
Root never be undefined, this unit test case requires **_parentRecord_** only, so others are passed as undefined